### ISSUE TITLE
TSIG canonicalization

### ISSUE THREAD

**Issue by matje** [miekg/dns#572]
When doing a query with TSIG

```
client := dns.Client{...}
msg := dns.Msg{...}
...
client.TsigSecret = map[string]string{"FooBar": "abcd"}
m.SetTsig(client.tsig, dns.HmacMD5, 300, time.Now().Unix())
r, _, err = client.Exchange(m, server)
```

I will get the error "dns: domain must be fully qualified" (it is actually documented that it must be). If I add a dot at the end of "FooBar" to make it a FQDN, I will get the error "dns: no secrets defined". 

This is because the configured name has uppercase characters, but the TSIG needs to be put in canonical wire format and so when a lookup is done in the TsigSecret map there is no match. If I lowercase the TSIG name to "foobar." it works fine.

To be honest I think this is unexpected. If a DNS library or server reads some TSIG configuration I expect it to be able to correctly parse a valid zone name. Meaning in this case "FooBar" is matched in a canonicalized way.

I think a change is justified:
1. Either update the documentation to say that the TSIG name should not only be FQDN, but also be presented in its canonical form.
2. I prefer to make the changes to convert the zonenames from TsigSecret to their canonical form such that it will match when looking up a TSIG secret given the TSIG owner name from the DNS packet. Not sure what the best way to do this, perhaps by having a method `setTSIG` that will do the canonicalization?
3. Something else.

What would be your preference?

**Comment by miekg** [comment]
There is no parsing going on as you just set things in a map, but making this more explicit in the docs seems like good plan. Took some RFC searching, but https://tools.ietf.org/html/rfc4034#section-6.2 actually defines "Canonical RR Form". (TIL)

**Comment by miekg** [comment]
the error returned should also include the name it used for the lookup to make things slightly more obvious

**Comment by matje** [comment]
Yes I know the RFC ad 4034 talks about wire format when defining canonical form, not string format.

**Comment by matje** [comment]
I made a PR to at least update the documentation.

https://github.com/miekg/dns/pull/574

**Comment by miekg** [comment]
closed via #574
